File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.py
from collections import deque, defaultdict

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None

class Solution:
    def markParents(self, root, parent_track):
        q = deque()
        q.append(root)
        while q:
            curr = q.popleft()
            if curr.left:
                parent_track[curr.left] = curr
                q.append(curr.left)
            if curr.right:
                parent_track[curr.right] = curr
                q.append(curr.right)

    def distanceK(self, root, target, k):
        parent_track = dict()
        self.markParents(root, parent_track)

        visited = set()
        q = deque()
        q.append(target)
        visited.add(target)
        curr_level = 0

        while q:
            if curr_level == k:
                break
            curr_level += 1
            for _ in range(len(q)):
                curr = q.popleft()
                if curr.left and curr.left not in visited:
                    q.append(curr.left)
                    visited.add(curr.left)
                if curr.right and curr.right not in visited:
                    q.append(curr.right)
                    visited.add(curr.right)
                if curr in parent_track and parent_track[curr] not in visited:
                    q.append(parent_track[curr])
                    visited.add(parent_track[curr])

        return [node.val for node in q]

File: 893-all-nodes-distance-k-in-binary-tree/all-nodes-distance-k-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    void markParents(TreeNode*root,unordered_map<TreeNode*,TreeNode*>&parent_track,TreeNode*target){
        queue<TreeNode*>q;
        q.push(root);
        while(!q.empty()){
            TreeNode*curr=q.front();
            q.pop();
            if(curr->left){
                parent_track[curr->left]=curr;
                q.push(curr->left);
            }
            if(curr->right){
                parent_track[curr->right]=curr;
                q.push(curr->right);
            }
        }
    }
    vector<int> distanceK(TreeNode* root, TreeNode* target, int k) {
       unordered_map <TreeNode*,TreeNode*> parent_track;
       markParents(root,parent_track,target);

       unordered_map<TreeNode*,bool>vis;
       vis[target]=true;
       queue<TreeNode*> q;
       q.push(target);
       int curr_level=0;
       while(!q.empty()){
        int size=q.size();
        if(curr_level==k)break;
        curr_level++;
        for(int i = 0 ; i < size ; i++){
            TreeNode*curr=q.front();
            q.pop();
            if(curr->left && !vis[curr->left]){
                q.push(curr->left);
                vis[curr->left]=true;
            }
            if(curr->right && !vis[curr->right]){
                q.push(curr->right);
                vis[curr->right]=true;
            }
            if(parent_track[curr] && !vis[parent_track[curr]]){
                q.push(parent_track[curr]);
                vis[parent_track[curr]]=true;
            }
        }
       }
       vector<int> result;
       while(!q.empty()){
         TreeNode*node=q.front();
         q.pop();
         result.push_back(node->val);
       }
       return result;
    }
};